tools/migrate_kv.ts

```diff
@@ -1,14 +1,140 @@
+    ...newVoteProps(),
+    item,
+    userLogin: voterUserRes.value.login,
+  };
+
+  const itemKey = ["items", vote.item.id];
+  const itemsByTimeKey = [
+    "items_by_time",
+    vote.item.createdAt.getTime(),
+    vote.item.id,
+  ];
+  const itemsByUserKey = ["items_by_user", vote.item.userLogin, vote.item.id];
+  const [itemRes, itemsByTimeRes, itemsByUserRes] = await kv.getMany([
+    itemKey,
+    itemsByTimeKey,
+    itemsByUserKey,
+  ]);
+  assertIsEntry(itemRes);
+  assertIsEntry(itemsByTimeRes);
+  assertIsEntry(itemsByUserRes);
+
+  const votesKey = ["votes", vote.id];
+  const votesByItemKey = ["votes_by_item", vote.item.id, vote.id];
+  const votesByUserKey = ["votes_by_user", vote.userLogin, vote.id];
+
+  const res = await kv.atomic()
+    .check(itemRes)
+    .check(itemsByTimeRes)
+    .check(itemsByUserRes)
+    .set(itemKey, vote.item)
+    .set(itemsByTimeKey, vote.item)
+    .set(itemsByUserKey, vote.item)
+    .set(votesKey, vote)
+    .set(votesByItemKey, vote)
+    .set(votesByUserKey, vote)
+    .commit();
+
+  if (!res.ok) throw new Error(`Failed to set vote: ${vote}`);
+}
+
+async function migrateItem(oldItem: OldItem) {
+  const userRes = await kv.get<User>(["users", oldItem.userId]);
+  assertIsEntry(userRes);
+
+  const item: Item = {
+    userLogin: userRes.value.login,
+    id: oldItem.id,
+    createdAt: oldItem.createdAt,
+    title: oldItem.title,
+    url: oldItem.url,
+    score: oldItem.score,
+  };
+
+  const itemsKey = ["items", item.id];
+  const itemsByTimeKey = ["items_by_time", item.createdAt.getTime(), item.id];
+  const itemsByUserKey = ["items_by_user", userRes.value.login, item.id];
+  const oldItemsByUserKey = ["items_by_user", oldItem.userId, item.id];
+
+  const res = await kv.atomic()
+    .set(itemsKey, item)
+    .set(itemsByTimeKey, item)
+    .set(itemsByUserKey, item)
+    .delete(oldItemsByUserKey)
+    .commit();
+
+  if (!res.ok) throw new Error(`Failed to migrate item: ${item}`);
+}
 
 export async function migrateKv() {
-  const iter = kv.list<Notification>({ prefix: ["notifications_by_time"] });
-  const promises = [];
-  for await (const entry of iter) {
-    promises.push(kv.delete(entry.key));
+  const promises1 = [];
+
+  // Notifications
+  const notificationsIter = kv.list<Notification>({
+    prefix: ["notifications"],
+  });
+  for await (const entry of notificationsIter) {
+    promises1.push(kv.delete(entry.key));
+  }
+
+  // Notifications by users
+  const notificationsByUsersIter = kv.list<Notification>({
+    prefix: ["notifications_by_user"],
+  });
+  for await (const entry of notificationsByUsersIter) {
+    promises1.push(kv.delete(entry.key));
+  }
+
+  // Votes
+  const votedItemsByUserIter = kv.list<OldItem>({
+    prefix: ["voted_items_by_user"],
+  });
+  for await (const entry of votedItemsByUserIter) {
+    promises1.push(migrateVote(entry.key[1].toString(), entry.value));
+  }
+  await Promise.all(promises1);
+
+  // Items
+  const promises2 = [];
```

Comment:
Both arrays of promises, `promises1` and `promises2`, must resolve sequentially. Vote migrations cover all items that have votes. Item migrations cover all remaining items that have no votes.

Using `Promise.all()` is optimal for performance.